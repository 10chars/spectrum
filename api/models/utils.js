// @flow
import { db } from './db';

export const NEW_DOCUMENTS = db
  .row('old_val')
  .eq(null)
  .and(db.not(db.row('new_val').eq(null)));

export const listenToNewDocumentsIn = (table: string, cb: Function) => {
  return (
    db
      .table(table)
      .changes({
        includeInitial: false,
      })
      // Filter to only include newly inserted messages in the changefeed
      .filter(NEW_DOCUMENTS)
      .run({ cursor: true })
      .then(cursor => {
        cursor.each((err, data) => {
          if (err) throw err;
          // Call the passed callback with the message directly
          cb(data.new_val);
        });
        return cursor;
      })
  );
};

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(db.row('lastSeen').during(db.now().sub(current), db.now()))
    .count()
    .default(0)
    .run();
};

export const getGrowth = async (
  table: string,
  range: Timeframe,
  field: string,
  filter: ?mixed
) => {
  const { current, previous } = parseRange(range);

  const [currentPeriodCount, previousPeriodCount] = await Promise.all([
    db
      .table(table)
      .filter(db.row(field).during(db.now().sub(current), db.now()))
      .filter(filter ? filter : '')
      .count()
      .run(),
    db
      .table(table)
      .filter(
        db.row(field).during(db.now().sub(previous), db.now().sub(current))
      )
      .filter(filter ? filter : '')
      .count()
      .run(),
  ]);
  const rate = (currentPeriodCount - previousPeriodCount) / previousPeriodCount;
  return {
    currentPeriodCount,
    previousPeriodCount,
    growthRate: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = (timeframe: number) => {
  return db
    .table('coreMetrics')
    .orderBy(db.desc('date'))
    .limit(timeframe)
    .orderBy('date')
    .run();
};

export const getActiveThreads = (range: number) => {
  return db
    .table('threads')
    .filter(db.row('lastActive').during(db.now().sub(range), db.now()))
    .filter(thread => db.not(thread.hasFields('deletedAt')))
    .group('communityId')
    .ungroup()
    .run();
};

const getCommunitiesWithMinimumMembers = (
  min: number = 2,
  communityIds: Array<string>
) => {
  return db
    .table('usersCommunities')
    .getAll(...communityIds, { index: 'communityId' })
    .group('communityId')
    .ungroup()
    .filter(row =>
      row('reduction')
        .count()
        .gt(min)
    )
    .map(row => row('group'))
    .run();
};

export const getAc = async (range: Timeframe) => {
  // constants
  const { current } = parseRange(range);
  const MIN_THREAD_COUNT = 1;

  // get threads posted in the range
  const threadsPostedInRange = await getActiveThreads(current);
  // returns an array of community ids
  const activeCommunitiesByThreads = threadsPostedInRange
    .filter(t => t.reduction.length > MIN_THREAD_COUNT)
    .map(t => t.group);

  // for each active community by thread count, only return communities with at least 2 members
  const activeCommunitiesByMember = await getCommunitiesWithMinimumMembers(
    2,
    activeCommunitiesByThreads
  );

  return activeCommunitiesByMember.length;
};

export const getAt = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('threads')
    .filter(db.row('lastActive').during(db.now().sub(current), db.now()))
    .count()
    .default(0)
    .run();
};
