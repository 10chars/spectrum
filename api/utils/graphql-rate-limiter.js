/* @flow
 *
 * GraphQL rate limiting validation rule
 *
 */
import type {
  OperationDefinitionNode,
  ValidationContext,
  Schema,
} from 'graphql';

const getFields = (schema: Schema, operation: OperationDefinitionNode) => {
  switch (operation.operation) {
    case 'query': {
      return schema.getQueryType().getFields();
    }
    case 'mutation': {
      return schema.getMutationType().getFields();
    }
    default: {
      return;
    }
  }
};

export default (options?: {}) => (context: ValidationContext) => {
  return {
    OperationDefinition: {
      enter: (operation: OperationDefinitionNode) => {
        const schema = context.getSchema();
        const fields = getFields(schema, operation);
        const field = operation.selectionSet.selections.find(
          ({ kind }) => kind === 'Field'
        );
        if (!fields || !field) return;

        const type = fields[field.name.value];
        if (!type) return;

        const rateLimitingDirective = type.astNode.directives.find(
          ({ name }) => name.value === 'rateLimit'
        );
        if (!rateLimitingDirective) return;
        console.log(rateLimitingDirective);
      },
    },
  };
};
