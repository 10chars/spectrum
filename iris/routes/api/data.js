// @flow
import { Router } from 'express';
import { db } from '../../models/db';
const dataRouter = Router();

// $FlowIssue
dataRouter.get('/coremetrics', async (req, res) => {
  console.log('getting core metrics');
  const json = await db.table('coreMetrics').run();
  res.send(JSON.stringify(json));
});

export default dataRouter;
