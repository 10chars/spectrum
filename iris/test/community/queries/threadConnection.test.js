// @flow
import { request } from '../../utils';
import { threads, communities } from 'shared/testing/data';

const community = communities[0];

it('should fetch a communities threads', async () => {
  const query = /* GraphQL */ `
    {
      community(id: "${community.id}") {
        threadConnection {
          edges {
            node {
              content {
                title
              }
            }
          }
        }
      }
    }
  `;

  expect.hasAssertions();
  const result = await request(query);

  expect(result).toMatchSnapshot();
});

it('should fetch a communities threads in lastActive order', async () => {
  const query = /* GraphQL */ `
    {
      community(id: "${community.id}") {
        threadConnection {
          edges {
            node {
              lastActive
              id
            }
          }
        }
      }
    }
  `;

  expect.hasAssertions();
  const result = await request(query);

  // Expect the manually by lastActive sorted result to equal the original result
  expect(
    result.data.community.threadConnection.edges.sort(
      (a, b) =>
        new Date(a.node.lastActive).getTime() - new Date(b.lastActive).getTime()
    )
  ).toEqual(result.data.community.threadConnection.edges);
});

describe('pagination', () => {
  it('should paginate threads', async () => {
    expect.hasAssertions();
    const query = /* GraphQL */ `
      {
        community(id: "${community.id}") {
          threadConnection(first: 1) {
            edges {
              cursor
              node {
                lastActive
                id
              }
            }
          }
        }
      }
    `;

    const firstThread = await request(query);

    expect(firstThread.data.community.threadConnection.edges.length).toEqual(1);
    expect(firstThread).toMatchSnapshot();

    // Construct a query with the cursor of the first thread
    const secondQuery = /* GraphQL */ `
      {
        community(id: "${community.id}") {
          threadConnection(first: 1, after: "${
            firstThread.data.community.threadConnection.edges[0].cursor
          }") {
            edges {
              cursor
              node {
                lastActive
                id
              }
            }
          }
        }
      }
    `;

    const secondThread = await request(secondQuery);

    expect(secondThread).not.toEqual(firstThread);
    expect(
      new Date(
        secondThread.data.community.threadConnection.edges[0].node.lastActive
      ).getTime()
    ).toBeLessThan(
      new Date(
        firstThread.data.community.threadConnection.edges[0].node.lastActive
      ).getTime()
    );
    expect(secondThread).toMatchSnapshot();
  });

  it('should be able to reverse paginate threads', async () => {
    expect.hasAssertions();
    const query = /* GraphQL */ `
      {
        community(id: "${community.id}") {
          threadConnection(last: 1) {
            edges {
              cursor
              node {
                lastActive
                id
              }
            }
          }
        }
      }
    `;

    const lastThread = await request(query);

    expect(lastThread.data.community.threadConnection.edges.length).toEqual(1);
    expect(lastThread).toMatchSnapshot();

    // Construct a query with the cursor of the last thread
    const secondQuery = /* GraphQL */ `
      {
        community(id: "${community.id}") {
          threadConnection(last: 1, before: "${
            lastThread.data.community.threadConnection.edges[0].cursor
          }") {
            edges {
              cursor
              node {
                lastActive
                id
              }
            }
          }
        }
      }
    `;

    const secondThread = await request(secondQuery);

    expect(secondThread).not.toEqual(lastThread);
    expect(
      new Date(
        secondThread.data.community.threadConnection.edges[0].node.lastActive
      ).getTime()
    ).toBeGreaterThan(
      new Date(
        lastThread.data.community.threadConnection.edges[0].node.lastActive
      ).getTime()
    );
    expect(secondThread).toMatchSnapshot();
  });
});
