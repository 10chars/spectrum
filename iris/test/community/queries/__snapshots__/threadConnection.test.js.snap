// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination should paginate threads 1`] = `
Object {
  "data": Object {
    "community": Object {
      "threadConnection": Object {
        "edges": Array [
          Object {
            "cursor": "ZjJlYjlkM2QtZWQwNS00OWFlLThmYzktOTFkMDIzMTRkNWE5LTE=",
            "node": Object {
              "id": "f2eb9d3d-ed05-49ae-8fc9-91d02314d5a9",
              "lastActive": "2016-12-31T23:00:00.002Z",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`pagination should paginate threads 2`] = `
Object {
  "data": Object {
    "community": Object {
      "threadConnection": Object {
        "edges": Array [
          Object {
            "cursor": "MTFlNzM2YjMtNTQ2NC00YmFiLWFjZmQtYmJkNDJjZGRjMWRkLTI=",
            "node": Object {
              "id": "11e736b3-5464-4bab-acfd-bbd42cddc1dd",
              "lastActive": "2016-12-31T23:00:00.001Z",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`should fetch a communities threads 1`] = `
Object {
  "data": Object {
    "community": Object {
      "threadConnection": Object {
        "edges": Array [
          Object {
            "node": Object {
              "content": Object {
                "title": "Yet another thread",
              },
            },
          },
          Object {
            "node": Object {
              "content": Object {
                "title": "Another thread",
              },
            },
          },
          Object {
            "node": Object {
              "content": Object {
                "title": "The first thread! ðŸŽ‰",
              },
            },
          },
        ],
      },
    },
  },
}
`;
